---
title: "Differential expression analysis comparison of subclones aggregated across samples"
params:
  seed: 42
  sce_process_fun: "/mnt/data/NGS/Projects/MCL-scrnaseq/MCL-PhanthomMenace/scrnaseq-analysis-nxf/bin/process_sce.R"
  pseudobulk_fun: "/mnt/data/NGS/Projects/MCL-scrnaseq/MCL-PhanthomMenace/DifferentialExpression/pseudobulk_minimal.r"
---

```{r setup}
seed <- params$seed
set.seed(seed)

library(tidyverse)
library(SingleCellExperiment)
library(scater)

library(AnnotationHub)
library(GenomicRanges)

library(ComplexHeatmap)
library(circlize)

library(GenVisR)

library(msigdbr)
library(clusterProfiler)

library(patchwork)

library(openxlsx)

sce_process_fun <- params$sce_process_fun
source(sce_process_fun)

pseudobulk_fun <- params$pseudobulk_fun
source(pseudobulk_fun)
```

# The dataset

```{r}
library(muscData)
sce <- Kang18_8vs8()

sce <- sce[, sce$multiplets == "singlet"]
sce <- sce[, !is.na(sce$cell)]

sce
```

```{r}
colData(sce) %>% 
  as.data.frame() %>% 
  glimpse()
```


```{r}
colData(sce) %>% 
  as.data.frame() %>% 
  remove_rownames() %>% 
  group_by(ind, stim) %>% 
  reframe(cells = n()) %>% 
  pivot_wider(names_from = stim, values_from = cells)
```

```{r}
colData(sce) %>% 
  as.data.frame() %>% 
  remove_rownames() %>% 
  group_by(stim, cell) %>% 
  reframe(cells = n()) %>% 
  pivot_wider(names_from = stim, values_from = cells)
```

```{r}
sce
```


```{r}
sce <- process_sce(
  sce,
  rowdata = "SYMBOL",
  HVF_method = "scran_mgv",
  n_HVGs = 2000,
  HVG_fdr_threshold = 0.1,
  exclude_HVGs = NULL,
  vars_to_regress = NULL,
  RunPCA_npcs = 40,
  pca_variance_threshold = 2,
  chosen_pcs = NULL,
  UMAP_min_dist = 0.3,
  UMAP_n_neighbors = 30,
  FindNeighbors_knn = 20,
  FindClusters_res = 0.5,
  seed = seed
)
```

```{r, fig.width=11, fig.height=5}
sce$ind <- factor(sce$ind)

plotUMAP(sce, color_by = "ind", other_fields = "stim") +
  facet_wrap( ~ stim) +
  theme_bw(base_size = 16)
```

```{r, fig.width=11, fig.height=5}
plotUMAP(sce, color_by = "cell", other_fields = "stim") +
  facet_wrap( ~ stim) +
  theme_bw(base_size = 16)
```

# Aggregation across one cell type

```{r}
sce_test <- sce[, sce$cell == "B cells"]
```

```{r}
table(sce_test$ind, sce_test$stim)
```

```{r}
# Ensure the correct comparison direction - first level is compared to the second
sce_test$stim <- factor(sce_test$stim, levels = c("ctrl", "stim"))

levels(sce_test$stim)
```

# Differential expression

```{r}
# create combined ID for proper within-sample aggregation
combined_id <- paste(sce$ind, sce$stim, sep = "_")
unique(combined_id)
```

```{r}
DESeq2_res <- RunPseudobulkMethod(
  raw.data = counts(sce), 
  normalized.data = logcounts(sce),
  individual = combined_id,
  group = sce$stim,
  test = "DESeq2",
  sum.or.mean = "sum"
  )
```

```{r}
pvalue_cutoff <- 0.01

DESeq2_res_sig <- DESeq2_res %>% 
  na.omit() %>% 
  arrange(desc(abs(logFC))) %>% 
  filter(
    padj < pvalue_cutoff
  )
DESeq2_res_sig
```

```{r}
DESeq2_res %>%
  na.omit() %>%
  ggplot(., aes(x = logFC, y = -log10(padj))) +
  geom_point() +
  theme_bw(base_size = 14) +
  labs(x = "log2 (Fold Change)", y = "-log10 (adjusted p value)") +
  geom_hline(
    yintercept = -log10(pvalue_cutoff),
    linetype = "dashed",
    color = "firebrick"
  ) +
  ggrepel::geom_text_repel(
    data = slice_head(DESeq2_res_sig, n = 10),
    aes(label = gene),
    color = "firebrick"
  )
```

```{r, fig.width=11, fig.height=5}
plotUMAP(sce, color_by = "CXCL10", other_fields = "stim") +
  facet_wrap( ~ stim) +
  theme_bw(base_size = 16)
```

# GSEA

```{r}
ranking <- DESeq2_res %>% 
  filter(!is.na(logFC)) %>% 
  arrange(., desc(logFC))

plot(
  x = seq_along(ranking$logFC),
  y = ranking$logFC,
  main = "Log2 FC ranking"
)
```

```{r}
gene_list <- ranking$logFC
names(gene_list) <- ranking$gene

gene_list <- sort(gene_list, decreasing = TRUE)

msigdbr_reactome_sets <- msigdbr(species = "Homo sapiens",
                        collection = "C2",
                        subcollection = "CP:REACTOME")

gsea_reactome_obj <- GSEA(
  geneList = gene_list,
  TERM2GENE = msigdbr_reactome_sets[, c("gs_name", "gene_symbol")],
  pvalueCutoff = 0.01,
  eps = 0,
  verbose = TRUE
)
```


```{r}
gsea_reactome_res <- gsea_reactome_obj@result %>%
  right_join(
    dplyr::select(msigdbr_reactome_sets, gs_name, gs_description) %>% distinct(),
    .,
    by = join_by("gs_name" == "ID")
  ) %>%
  dplyr::select(-Description) %>%
  filter(!is.na(setSize)) %>%
  mutate(
    core_count = sapply(strsplit(core_enrichment, "/"), length),
    gene_ratio = core_count / setSize
  ) %>%
  arrange(qvalue)
gsea_reactome_res
```

# Sesson info

```{r}
sessionInfo()
```

